modal/_type_manager.py

```diff
@@ -0,0 +1,198 @@
+# Copyright Modal Labs 2025
+import typing
+from typing import Any
+
+import typing_extensions
+
+from modal.exception import InvalidError
+from modal_proto import api_pb2
+
+
+class ParameterProtoSerde(typing.Protocol):
+    def encode(self, value: Any, name: typing.Optional[str] = None) -> api_pb2.ClassParameterValue:
+        pass
+
+    def decode(self, proto: api_pb2.ClassParameterValue) -> Any:
+        pass
+
+
+class ProtoParameterSerdeRegistry:
+    _py_base_type_to_serde: dict[type, ParameterProtoSerde]
+    _proto_type_to_serde: dict["api_pb2.ParameterType.ValueType", ParameterProtoSerde]
+
+    def __init__(self):
+        self._py_base_type_to_serde = {}
+        self._proto_type_to_serde = {}
+
+    def register_encoder(self, python_base_type: type) -> typing.Callable[[ParameterProtoSerde], ParameterProtoSerde]:
+        def deco(ph: ParameterProtoSerde) -> ParameterProtoSerde:
+            if python_base_type in self._py_base_type_to_serde:
+                raise ValueError("Can't register the same encoder type twice")
+            self._py_base_type_to_serde[python_base_type] = ph
+            return ph
+
+        return deco
+
+    def register_decoder(self, enum_type_value: "api_pb2.ParameterType.ValueType"):
+        def deco(ph: ParameterProtoSerde) -> ParameterProtoSerde:
+            if enum_type_value in self._proto_type_to_serde:
+                raise ValueError("Can't register the same decoder type twice")
+            self._proto_type_to_serde[enum_type_value] = ph
+            return ph
+
+        return deco
+
+    def get_encoder(self, python_base_type: type) -> ParameterProtoSerde:
+        try:
+            return self._py_base_type_to_serde[python_base_type]
+        except InvalidError:
+            raise KeyError(f"No decoder implemented for python type {python_base_type.__name__}")
+
+    def encode(self, python_value: type) -> api_pb2.ClassParameterValue:
+        return self.get_encoder(type(python_value)).encode(python_value)
```

Comment:
Are users of this class meant to interact just with `encode`/`decode` or also with `get_encoder`/`get_decoder` directly?

Reply:
Users being us. In most cases it will be enough to just use encode/decode, but there are e couple of places where the getters are used. I'll see if I can refactor those cases that might be cleaner

Reply:
I refactored to put the validation logic in the `ProtoParameterSerdeRegistry` instead, so now the get_decoder and get_encoder are both private there